lib/Generate.js

```diff
@@ -59,110 +61,51 @@ function addKernelTestChecking(content) {
   );
 }
 
-function generateElmJson(
-  projectRootDir /*: string */,
-  generatedCodeDir /*: string */,
-  hasBeenGivenCustomGlobs /*: boolean */,
-  elmJsonPath /*: string */,
-  projectElmJson /*: any */
-) /*: [string, Array<string>] */ {
-  const testRootDir = Compile.getTestRootDir(projectRootDir);
-  const generatedSrc = path.join(generatedCodeDir, 'src');
-
-  var isPackageProject = projectElmJson.type === 'package';
-
-  // if we were given file globs, we don't need to check the tests/ directory exists
-  // this is only for elm applications, which people may need to introduce slowly into their apps
-  // for packages, we stick with the existing behaviour and assume tests are in tests/ so do the check always
-  const needToCareAboutTestsDir =
-    hasBeenGivenCustomGlobs === false || isPackageProject === true;
-
-  // we add the tests dir as a source if:
-  // - we decided above we need to care about it
-  // - if it exists on disk: this supports the case where we have all our tests in tests/ but
-  //   want to pass a glob to only run one of the test files
-  const shouldAddTestsDirAsSource =
-    needToCareAboutTestsDir ||
-    fs.existsSync(path.join(projectRootDir, 'tests'));
-
-  if (needToCareAboutTestsDir) {
-    if (!fs.existsSync(testRootDir)) {
-      console.error(
-        'Error: ' +
-          testRootDir +
-          ' does not exist. Please create a tests/ directory in your project root!'
-      );
-      process.exit(1);
-    }
-
-    if (!fs.lstatSync(testRootDir).isDirectory()) {
-      console.error(
-        'Error: ' +
-          testRootDir +
-          ' exists, but it is not a directory. Please create a tests/ directory in your project root!'
-      );
-      process.exit(1);
-    }
-  }
+function getGeneratedSrcDir(generatedCodeDir /*: string */) /*: string */ {
+  return path.join(generatedCodeDir, 'src');
+}
+
+function generateElmJson(project /*: typeof Project.Project */) /*: void */ {
```

Comment:
This function was doing a lot of calculations that other parts of the code needs as well, so I moved those things into Project.js. Now, this function instead just takes a `project: Project` and could be simplified.